forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceBindingCallback.kt

```diff
@@ -0,0 +1,160 @@
+ * of the MIT license. See the LICENSE file for details.
+ */
+package org.forgerock.android.auth.callback
+
+import android.content.Context
+import androidx.biometric.BiometricPrompt
+import org.forgerock.android.auth.FRListener
+import org.forgerock.android.auth.Listener
+import org.forgerock.android.auth.Logger
+import org.forgerock.android.auth.biometric.BiometricAuthCompletionHandler
+import org.forgerock.android.auth.devicebind.*
+import org.json.JSONObject
+import java.util.*
+import java.util.concurrent.TimeUnit
+
+open class DeviceBindingCallback: AbstractCallback {
+
+    @JvmOverloads constructor(jsonObject: JSONObject,
+                              index: Int): super(jsonObject, index)
+
+    @JvmOverloads constructor(): super()
+
+    private lateinit var userId: String
+    private lateinit var challenge: String
+    private lateinit var deviceBindingAuthenticationType: DeviceBindingAuthenticationType
+    private lateinit var title: String
+    private lateinit var subtitle: String
+    private lateinit var description: String
+    private var timeout: Int? = null
+    private val tag = DeviceBindingCallback::class.java.simpleName
+
+    override fun setAttribute(name: String, value: Any) = when (name) {
+        "userId" ->  userId = value as String
+        "challenge" -> challenge = value as String
+        "authenticationType" -> {
+            deviceBindingAuthenticationType = (value as? String)?.let {
+                DeviceBindingAuthenticationType.valueOf(it)
+            } ?: DeviceBindingAuthenticationType.NONE
+        }
+        "title" -> title = value as? String ?: ""
+        "subtitle" -> subtitle = value as? String ?: ""
+        "description" -> description = value as? String ?: ""
+        "timeout" -> timeout = value as? Int
+        else -> {}
+    }
+
+    override fun getType(): String {
+        return "DeviceBindingCallback"
+    }
+
+    fun setJws(value: String?) {
+        super.setValue(value, 0)
+    }
+
+    fun setDeviceName(value: String?) {
+        super.setValue(value, 1)
+    }
+
+    fun setClientError(value: String?) {
+        super.setValue(value, 2)
+    }
+
+    fun bind(context: Context,
+             listener: FRListener<Void>) {
+        execute(context, listener)
+    }
+
+     internal fun execute(context: Context,
+                         listener: FRListener<Void>,
+                         keyAware: KeyAwareInterface = KeyAware(userId),
+                         biometric: BiometricInterface? = BiometricUtil(title, subtitle, description, deviceBindAuthenticationType = deviceBindingAuthenticationType),
+                         encryptedPreference: PreferenceInterface = PreferenceUtil(context)) {
+
+        val keypair: KeyPair
+        // If the authentication type is none or the biometric supported then create keypair. if not throw exception
+        if(deviceBindingAuthenticationType == DeviceBindingAuthenticationType.NONE || isSupported(biometric)) {
```

Comment:
As discussed we may able to change it to template method design pattern, or using sealed classes.